### PR TITLE
Retain stack frame in debug mode

### PR DESCRIPTION
**Pull request by jmileson**
Make it easier to debug unexpected exceptions in debug mode.

### CODE REVIEW

File: flask_rest_jsonapi/decorators.py [hunk 1]
```diff
@@ -80,7 +84,7 @@ def wrapper(*args, **kwargs):
                                  headers)
         except Exception as e:
             if current_app.config['DEBUG'] is True:
-                raise e
+                six.reraise(*sys.exc_info())
```
Comment (kornholi) [inline flask_rest_jsonapi/decorators.py]: does a bare `raise` not work here?

Comment (jmileson) [reply flask_rest_jsonapi/decorators.py]: It works, but you don't preserve the context of the stack trace, so it's difficult to determine where the exception was thrown.  With `raise e` you only see the exception in the context of the decorator like:

```
Traceback (most recent call last):
...
"test_v2.runfiles/requirements_py2_pypi__Flask_REST_JSONAPI_0_29_0/flask_rest_jsonapi/decorators.py", line 87, in wrapper
    raise e
Exception: failed
```

While with `six.reraise(*sys.exc_info()` you see the original context that raised the exception:

```
Traceback (most recent call last):
...
  File "test_v2.runfiles/requirements_py2_pypi__marshmallow_2_18_0/marshmallow/marshalling.py", line 283, in <lambda>
    data
  File "test_v2.runfiles/requirements_py2_pypi__marshmallow_2_18_0/marshmallow/fields.py", line 265, in deserialize
    output = self._deserialize(value, attr, data)
  File "test_v2.runfiles/__main__/fields.py", line 62, in _deserialize
    raise Exception("failed")
Exception: failed
```

Comment (kornholi) [reply flask_rest_jsonapi/decorators.py]: https://docs.python.org/2.7/reference/simple_stmts.html#raise

Comment (jmileson) [reply flask_rest_jsonapi/decorators.py]: `raise` has a different syntax in python3

Comment (kornholi) [reply flask_rest_jsonapi/decorators.py]: what's different about it in py3 that wouldn't work here?

Comment (jmileson) [reply flask_rest_jsonapi/decorators.py]: Specifically, the `raise <type>, <value>, <traceback>` form doesn't exist any more in favor of `raise <value>.with_traceback(<traceback>)`. 

Python3 is better about retaining the correct info anyway, so given the following function:

In python3:
``` python
def raiser():
    raise Exception("foo")
```
These two behave basically the same:
``` python
try:
    raiser()
except Exception as e:
    raise e.with_traceback(sys.exc_info()[2])

try:
    raiser()
except Exception as e:
    raise e
```

The traceback for both of these implementations clearly shows where the exception originated from:
```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-4-fefcf4ed47ea> in <module>
      2     raiser()
      3 except Exception as e:
----> 4     raise e.with_traceback(sys.exc_info()[2])
      5

<ipython-input-4-fefcf4ed47ea> in <module>
      1 try:
----> 2     raiser()
      3 except Exception as e:
      4     raise e.with_traceback(sys.exc_info()[2])
      5

<ipython-input-3-bb248eb5a561> in raiser()
      1 def raiser():
----> 2     raise Exception("foo")
      3

Exception: foo
```

In python 2, the `raise e` form drops the original traceback
``` python
try:
    raiser()
except Exception as e:
    raise e
```

So you can't see any info in the traceback about what line originated the exception.
```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-14-5981f9f962cd> in <module>()
      2     raiser()
      3 except Exception as e:
----> 4     raise e

Exception: foo
```

In order to get that behavior in python2, you need to use this form:
``` python
try:
    raiser()
except Exception as e:
    type_, value, tb = sys.exc_info()
    raise type_, value, tb
```

You get a better traceback, though less complete than the python3 version:
```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-13-7a5b66c805f2> in <module>()
      1 try:
----> 2     raiser()
      3 except Exception as e:
      4     type_, value, tb = sys.exc_info()
      5     raise type_, value, tb

<ipython-input-9-8127c176ebf6> in raiser()
      1 def raiser():
----> 2     raise Exception("foo")

Exception: foo
```

`with_traceback` doesn't exist in python2, so you need the 3-tuple form to get this behavior.

**Comment by nikhaldi** [comment]
Any chance this can be merged? This improves fidelity a lot when running flask locally and in unit tests. Without a complete stacktrace it can be really hard to track down a problem.

**Pull request merged**